src/frame.rs

```diff
@@ -602,20 +602,13 @@ pub type FrameResult = Result<Option<Block>>;
 /// A function to expand the length of a buffer, or replace the buffer altogether,
 /// so it can hold at least `new_len` elements. The contents of the buffer can
 /// be anything, it is assumed they will be overwritten anyway.
-///
-/// To use this function safely, the caller must overwrite all `new_len` bytes.
-unsafe fn ensure_buffer_len(mut buffer: Vec<i32>, new_len: usize) -> Vec<i32> {
+fn ensure_buffer_len(mut buffer: Vec<i32>, new_len: usize) -> Vec<i32> {
     if buffer.len() < new_len {
         // Previous data will be overwritten, so instead of resizing the
         // vector if it is too small, we might as well allocate a new one.
         if buffer.capacity() < new_len {
-            buffer = Vec::with_capacity(new_len);
+            buffer = vec![0; new_len];
         }
-
-        // We are going to fill the buffer anyway, so there is no point in
-        // initializing it with default values. This does mean that there could
-        // be garbage in the buffer, therefore this function is unsafe.
-        buffer.set_len(new_len);
```

Comment:
This now misses the case where the buffer has enough capacity, but not enough elements. That can cause a buffer to be returned whose `len()` is not `new_len`.

Reply:
The line `buffer = vec![0; new_len];` always returns a vector with length `new_len`. It desugars to `vec.from_elem()`,  you can check out the implementation [here](https://doc.rust-lang.org/src/alloc/vec.rs.html#1580). Or am I missing the bigger picture?

Reply:
This fails:
```rust
#[test]
fn ensure_buffer_len_returns_buffer_with_new_len() {
    let buffer = Vec::with_capacity(1);
    assert_eq!(ensure_buffer_len(buffer, 1).len(), 1);
}
```

Reply:
Oooh, good catch. Thanks a lot!

I've fixed this bug by adding `buffer.resize(new_len, 0);` in case the capacity is sufficient but length is not. Now the worst case is going to be a very big capacity up front, then alternating it between very small and very large, but no more than the initial allocation.

I've followed up on your comments on the differential fuzzing harness and checked how slow `resize()` really is. Turns out the performance anomaly is only observable under Address Sanitizer; it's as fast as `vec!` macro with non-zero values or `repeat(x).take(len).collect()`. I can share the test harness if you're interested.